Option to run Django tests with manage.py

vindarel opened the pull request:
_"…setting the project root accordingly,
in order not to set the environnement variable DJANGO_SETTINGS_MODULE
for every project (but instead change of virtual env)."_

This PR does not change elpy's default behavior.

It adds a variable to tell elpy to run django tests with `manage.py` instead of `django-admin.py`. 
The thing is that we can't just set `elpy-test-django-runner-command` to our manage.py command because `elpy-test-at-point` needs to return the project root, not the library root.

Benefits: it allows a user to run tests quicker, with no venv or environment variable configuration. We just have to change of virtual env, which is easily done with other packages ([virtualenvwrapper.el](https://github.com/porterjamesj/virtualenvwrapper.el) and others). 

Because from my experience, I wasn't able to run django tests when I tried elpy. Now, I can and I hope it's easier for others too.

See https://github.com/jorgenschaefer/elpy/issues/931

jorgenschaefer commented:
Hello, and thanks for the contribution – especially improving the documentation at the same time is excellent!

The changes look basically ok to me. A few comments, though:

You are using `elpy-project-find-projectile-root` directly, which creates a hard projectile dependency. There is no need for this as far as I can see, you can just use `elpy-project-root`.

Or even better yet, you could use `locate-dominating-file` and search for a `manage.py` file directly.

Do you see a problem with that?

vindarel commented:
Rich idea, I added locate-dominating-file (nice method !) with elpy-project-root, expanding the file name just to be sure.

Indeed for the doc :)

(I rebased the PR)

jorgenschaefer commented:
Heh, that was fast :-) I'm afraid this won't work – the first argument to `locate-dominating-file` is the file name, the second is the directory. It also returns a file name, not a directory name.

You should use `locate-dominating-file` in elpy-test-django-runner to find the executable to run, not to change any top directory or anything.

vindarel commented:
I pushed the change with elpy-project-root.

I'm afraid I don't understand the other suggestion then.
Maybe can you comment on the specific line ?
(to me it would be `locate-dominating-file start-directory file-name-to-look-for`, which returns `~/project/`)

jorgenschaefer commented on elpy.el:
```diff
@@ -1989,7 +2001,9 @@ directory is not nil."
-    (let* ((top (elpy-library-root))
+    (let* ((top (if elpy-test-django-with-manage
+                  (elpy-project-root)
+                 (elpy-library-root)))
```
You should not have to change this function at all.

jorgenschaefer commented on elpy.el:
```diff
@@ -2050,14 +2064,17 @@ This requires Python 2.7 or later."
+             (append (if elpy-test-django-with-manage
+                         elpy-test-django-runner-manage-command
+                       elpy-test-django-runner-command)
                      (list (if test
```
Instead, use `locate-dominating-file` to get the full path to `manage.py` to use instead of `elpy-test-django-runner-command`.

vindarel replied:
So ! I don't clearly see what you suggest. I tried:
```lisp 
(list (concat 
    (expand-file-name (locate-dominating-file (elpy-project-root) "manage.py"))
    ;; returns a directory path, so I must append the file name again.
    ;; This snippet will get more verbose for the need of using the variable `elpy-test-django-runner-manage-command`. 
     "manage.py"))
```
the final command is:

    /home/.../parentdir/projectname/manage.py projectname.myapp.tests-dir.etc

the projectname in the last part is wrong. I suspect it's there because the `top` variable is at `parentdir`.

That's where I am. Anyway, I feel this is a lot of machinery, not necessarily cleaner than keeping the first "if" above, in `elpy-test-at-point` :]

jorgenschaefer replied:
`(concat (locate-dominating-file (buffer-file-name) "manage.py") "manage.py")` will give you the full path to a `manage.py` which you can use as the command to run. This is a lot cleaner than your solution, as it will work in a lot more situations. Please also think about what "the command to run" means, as that will simplify the rest of your code.

vindarel commented:
Ahah ! yes ?

jorgenschaefer commented on elpy.el:
```diff
@@ -2050,14 +2063,19 @@ This requires Python 2.7 or later."
+             (append (if elpy-test-django-with-manage
+                         elpy-test-django-runner-manage-command
+                       elpy-test-django-runner-command)
                      (list (if test
```
- The _directory_ in which to run `elpy-test-run` should *probably* be the directory of `manage.py`, no?
- The _command_ to run should be the full path to manage.py

vindarel commented:
Hi ! So, I know I didn't rebase the last two commits, it's on purpose for your review. Now it looks to me we are closer to what you ask: the code is changed in only one place. Hope it's good. It's working though :]

jorgenschaefer commented on elpy.el:
```diff
@@ -315,6 +315,18 @@ edited instead. Setting this variable to nil disables this feature."
   :group 'elpy)
 
+(defcustom elpy-test-django-runner-manage-command '("manage.py" "test"
+                                             "--noinput")
```
Why is this on a new line? And if it is, why is it not indented as per normal Emacs Lisp rules (just use `TAB`)?

vindarel replied:
sure, because of a search&replace.

jorgenschaefer commented on elpy.el:
```diff
@@ -315,6 +315,18 @@ edited instead. Setting this variable to nil disables this feature."
+
+(defcustom elpy-test-django-with-manage nil
+  "Set to nil, elpy will use `elpy-test-django-runner-command',
+   set to t elpy will use `elpy-test-django-runner-manage-command' and set the project root accordingly."
```
No spaces at the beginning of the line here.

jorgenschaefer commented on elpy.el:
```diff
@@ -1989,7 +2001,8 @@ directory is not nil."
         (list (elpy-library-root) nil nil nil))
-    (let* ((top (elpy-library-root))
+    (let* ((top
+            (elpy-library-root))
```
No need to add a line break here.

jorgenschaefer commented on elpy.el:
```diff
@@ -2050,17 +2063,27 @@ This requires Python 2.7 or later."
 (defun elpy-test-django-runner (top file module test)
-  "Test the project using the Django discover runner.
+  "Test the project using the Django discover runner,
+   or with manage.py if elpy-test-django-with-manage is true.
```
No need for spaces at the beginning of the line here.

jorgenschaefer commented on elpy.el:
```diff
@@ -2050,17 +2063,27 @@ This requires Python 2.7 or later."
+              (if elpy-test-django-with-manage
+                         (append
+                          (list (concat
+                           (expand-file-name (locate-dominating-file (elpy-project-root) (car elpy-test-django-runner-manage-command)))
```
On the other hand, this line could really be split up on multiple lines to make it clearer what is happening here. (General rule of thumb, a line break can be useful after a number of consecutive closing parens).

jorgenschaefer commented:
Thank you for your contribution!

vindarel commented:
Awesome, thanks to you too :)
